### PR TITLE
initialization project

### PR DESCRIPTION
**Pull request by forbrock**

### CODE REVIEW

File: src/main/java/com/jdbc/dao/UserDao.java [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+            } else {
+                throw new SQLException();
+            }
+        } catch (SQLException e) {
+            System.err.println("Failed to connect");
+            e.printStackTrace();
+        }
+    }
+
+    public ArrayList<User> getUsers() {
```
Comment (grafgeest) [inline src/main/java/com/jdbc/dao/UserDao.java]: Use List instead

**Review by grafgeest — CHANGES_REQUESTED** [review]

File: .gitignore [hunk 2]
```diff
@@ -0,0 +1,6 @@
+.idea/*
+target/*
+*.class
+Users.iml
```
Comment (grafgeest) [inline .gitignore]: Use *.iml

File: src/main/java/com/jdbc/App.java [hunk 3]
```diff
@@ -0,0 +1,21 @@
+package com.jdbc;
+
+import com.jdbc.dao.UserDao;
+
+import java.sql.*;
+
+public class App {
+    public static void main(String[] args) {
+        UserDao userDao = new UserDao();
+        Connection connection = userDao.getConnection();
```
Comment (grafgeest) [inline src/main/java/com/jdbc/App.java]: connection should be closed during finding user

Comment (forbrock) [reply src/main/java/com/jdbc/App.java]: this feature will be fixed with new class and connection pool

File: .gitignore (hunk 2 above)
Comment (forbrock) [reply .gitignore]: accepted

File: src/main/java/com/jdbc/dao/Dao.java [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+
+import com.jdbc.model.User;
+
+import java.sql.SQLException;
+import java.util.List;
+
+public interface Dao {
+    boolean save(User userObject) throws SQLException;
+    User find(SearchOption searchOption, Object criterion) throws SQLException;
+    List<User> findAll() throws SQLException;
```
Comment (grafgeest) [inline src/main/java/com/jdbc/dao/Dao.java]: Use generics List <T>

File: src/main/java/com/jdbc/dao/Dao.java [hunk 5]
```diff
@@ -0,0 +1,14 @@
+package com.jdbc.dao;
+
+import com.jdbc.model.User;
+
+import java.sql.SQLException;
+import java.util.List;
+
+public interface Dao {
+    boolean save(User userObject) throws SQLException;
```
Comment (grafgeest) [inline src/main/java/com/jdbc/dao/Dao.java]: Use generics save(T obj)

File: src/main/java/com/jdbc/dao/UserDao.java [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+import java.time.LocalDate;
+import java.util.ArrayList;
+import java.util.List;
+
+public class UserDao implements Dao {
+    private static final Logger logger = LoggerFactory.getLogger(UserDao.class);
+    DataSource dataSource;
+
+    public UserDao(DBType type) {
+        DBCPDataSourceFactory dbcpFactory = new DBCPDataSourceFactory();
```
Comment (grafgeest) [inline src/main/java/com/jdbc/dao/UserDao.java]: Use static factory method

File: src/main/java/com/jdbc/dao/UserDao.java [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+                String name = resultSet.getString("name");
+                LocalDate birth = resultSet.getObject("birth", LocalDate.class);
+                String email = resultSet.getString("email");
+
+                user = new User(id, name, birth, email);
+            }
+        } catch (SQLException e) {
+            logger.error("Something went wrong", e);
+        } catch (NullPointerException e) {
+            logger.error(e.getMessage(), e);
```
Comment (grafgeest) [inline src/main/java/com/jdbc/dao/UserDao.java]: Use Optional to avoid NPE

File: src/main/java/com/jdbc/dao/UserDao.java [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+                return String.format(selectUser, "birth", criterion);
+            case BY_EMAIL:
+                return String.format(selectUser, "email", criterion);
+            default:
+                logger.error("Wrong choice during query creation");
+                throw new IllegalArgumentException();
+        }
+    }
+
+    public Connection getConnection() {
```
Comment (grafgeest) [inline src/main/java/com/jdbc/dao/UserDao.java]: Are you going to duplicate this logic in all DAO?

File: src/main/java/com/jdbc/dao/factory/DBCPDataSourceFactory.java [hunk 9]
```diff
@@ -0,0 +1,14 @@
+package com.jdbc.dao.factory;
+
+public class DBCPDataSourceFactory {
+
+    public DataSource getDBCPObject(DBType type) {
+        if (type == null) throw new IllegalArgumentException("Wrong database type: " + null);
```
Comment (grafgeest) [inline src/main/java/com/jdbc/dao/factory/DBCPDataSourceFactory.java]: Use Optional

File: src/main/java/com/jdbc/model/User.java [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+
+import java.time.format.DateTimeFormatter;
+import java.time.LocalDate;
+
+public class User {
+    private int id;
+    private String name;
+    private LocalDate birth;
+    private String email;
+    DateTimeFormatter formatter = DateTimeFormatter.ofPattern("d/MM/yyyy");
```
Comment (grafgeest) [inline src/main/java/com/jdbc/model/User.java]: private?

File: src/main/logs/AppLog.log [hunk 11]
```diff
@@ -0,0 +1,5 @@
+2020-09-17 22:43:29 [main] ERROR com.jdbc.dao.UserDao - null
```
Comment (grafgeest) [inline src/main/logs/AppLog.log]: I suppose this file was committed by mistake

File: src/test/java/com/jdbc/UserDaoTest.java [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+import java.util.List;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+public class UserDaoTest {
+    private static final Logger logger = LoggerFactory.getLogger(UserDaoTest.class);
+    private static UserDao userDao = new UserDao(DBType.H2_INMEMORY_DB);
+    DateTimeFormatter formatter = DateTimeFormatter.ofPattern("d/MM/yyyy");
+
+    @BeforeEach
```
Comment (grafgeest) [inline src/test/java/com/jdbc/UserDaoTest.java]: @AfterEach?

File: README.md [hunk 13]
```diff
@@ -1 +1,12 @@
-
+# Users
+CRUD implementation
+How to use methods:
+- _save(userObject)_, takes user object and saves it to the database, the id is adds automatically;
```
Comment (Besandr) [inline README.md:4]: Readme instructions should be used for describing how to install and use the application (not a Class). You shouldn't confuse javadocs with Readme.

File: src/main/java/com/jdbc/dao/UserDao.java [hunk 14]
```diff
@@ -0,0 +1,154 @@
...
+            connection = dataSource.getConnection();
+        } catch (SQLException e) {
+            logger.error("Connection isn't established", e);
+        }
+        return connection;
+    }
+
+    private User queryExecuteHelper(String query) {
+        User user = null;
+        try (Connection conn = getConnection();
```
Comment (grafgeest) [inline src/main/java/com/jdbc/dao/UserDao.java]: Connection conn = dataSource.getConnection()

File: src/main/java/com/jdbc/dao/UserDao.java [hunk 15]
```diff
@@ -0,0 +1,154 @@
...
+
+import javax.sql.DataSource;
+import java.sql.*;
+import java.time.LocalDate;
+import java.util.ArrayList;
+import java.util.List;
+
+public class UserDao implements Dao<User> {
+    private static final Logger logger = LoggerFactory.getLogger(UserDao.class);
+    DataSource dataSource;
```
Comment (grafgeest) [inline src/main/java/com/jdbc/dao/UserDao.java]: private?

File: src/main/java/com/jdbc/App.java [hunk 16]
```diff
@@ -0,0 +1,22 @@
...
+import com.jdbc.dao.factory.DBCPDataSourceFactory;
+import com.jdbc.dao.factory.DBType;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.sql.DataSource;
+
+public class App {
+    private static final Logger logger = LoggerFactory.getLogger(App.class);
+    private static DataSource dataSource = DBCPDataSourceFactory.getDataSource(DBType.POSTGRESQL);
```
Comment (grafgeest) [inline src/main/java/com/jdbc/App.java]: put it in main method

File: src/main/java/com/jdbc/dao/UserDao.java (hunk 14 above)
Comment (forbrock) [reply src/main/java/com/jdbc/dao/UserDao.java]: fixed

File: src/main/logs/AppLog.log (hunk 11 above)
Comment (forbrock) [reply src/main/logs/AppLog.log]: exactly

File: src/main/java/com/jdbc/dao/factory/DBCPDataSourceFactory.java (hunk 9 above)
Comment (forbrock) [reply src/main/java/com/jdbc/dao/factory/DBCPDataSourceFactory.java]: this check is redundant, and was deleted.

**Pull request merged**